### DISCUSSION TITLE
Slicing from a Compressed HDU

### DISCUSSION THREAD

**Discussion by at88mph** [nom-tam-fits/nom-tam-fits#344] [Q&A]
We at the CADC have a web enabled cutout service.  The cutout pixel range is specified in CFITSIO style:
`[1][20:50,60:100]`

In order to slice from a `CompressedImageHDU`, I would need to decompress it, probably with the `asImageHDU()` method.  That won't work well on the cloud with a 500MB file that would expand to Gigabytes, as the JVM would run out of memory quickly and crash.  I'm wondering what my options are with `nom-tam-fits`.

I _think_ I need to run tile by tile to see if my input Pixel ranges overlap, but I'm not sure.  Has anyone done anything similar?  Is there a way to calculate from the Header what row/tile I need to start extracting?  Thank you for any input.

**Comment by attipaci** [comment]
Hi!

That's a good question. In my mind the whole point of tile compression is exactly so that you can access individual image tiles without having to uncompress the entire image. (Although as far as I can tell Ritchie only implemented the decompression of the entire image). The tiling ought to be pretty easy to figure out. Basically the tile size is set by the `ZTILEn`keywords (which are analogous to `NAXISn`), and the tiles are then ordered...

Maybe, it would make sense to add some methods to retrieve select rectangles (n-dimensional cubes) from tile compressed images easier. E.g., something like `ImageData getImageArea(int[] offset, int[] size)`, which probably would rely on a lower-level `ImageData getImageTile(int index)` providing individual decompressed tiles, and a `int getTileIndex(int... pixelIndex)` providing the mapping of pixel locations to tile indixes. (And, the `getImageArea()` method above could also be implemented for regular non-compressed images also).

We could even complement that with a `setImageArea(ImageData patch, int... offset)`, for both compressed and uncompressed  images, which in turn could call a lower-level `setImageTile(int index, ImageData tile)`...

Would you fancy taking a shot at it? I'd be happy to provide feedback. We could target that for 1.18 (around 6/15/2023)...

Otherwise, I could put in on my docket, but I'm not sure when I'll really find the time to work on it (although there is some chance that I could get around to doing it for 1.18...)

**Comment by at88mph** [comment]
Thanks @attipaci ,

I would contribute this back, sure.  Do you know if decompressing a single row should work like this?

```java
PrimitiveType<Buffer> primitiveBufferType = ...
ICompressorControl compressController = CompressorProvider.findCompressorControl(null, "RICE_1",
                                                        primitiveBufferType.primitiveClass());
ICompressOption option = compressorControl.option();
int blockSize = ...
int bytePix = ...
if (option instanceof RiceCompressOption) {
    ((RiceCompressOption) option).setBlockSize(blockSize);
    ((RiceCompressOption) option).setBytePix(bytePix);
} else if (option instanceof RiceQuantizeCompressOption) {
    ((RiceQuantizeCompressOption) option).getRiceCompressOption().setBlockSize(blockSize);
    ((RiceQuantizeCompressOption) option).getRiceCompressOption().setBytePix(bytePix);
}

int tileWidth = ... // ZTILE1 or ZNAXIS1
int tileHeight = ... // ZTILE2 or 1
option.setTileHeight(tileHeight).setTileWidth(tileWidth);
option.getCompressionParameters().initializeColumns(4);

CompressedImageData compressedData = ...
while(moreRowsToRead) {
    int rowIndex = ...
    ByteBuffer compressed = ByteBuffer.wrap(compressedData.row(rowIndex));
    compressed.rewind();

    Buffer tileBuffer = primitiveBufferType.newBuffer(tileSize);
    tileBuffer.rewind();

    compressorControl.decompress(compressed, tileBuffer, option);
    read(tileBuffer);  // These float values seem to be incorrect.
    rowIndex++;
}
```

I'm comparing values from CFITSIO, which produce a recognizable sub image, but I think either my approach is incorrect, or the decompression isn't quite right.  Does it look valid?

**Comment by attipaci** [comment]
Hi @at88mph,

I wonder if this part now works more as expected with all the compression-related fixes and improvements since this discussion took place. Let me know if there is more to be done, or if we can close this discussion.
